bpf/lib/metrics.h

```diff
@@ -0,0 +1,62 @@
+/*
+ *  Copyright (C) 2018 Authors of Cilium
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+/*
+ * Data metrics collection functions
+ *
+ */
+
+#ifndef __LIB_METRICS__
+#define __LIB_METRICS__
+
+
+#include "common.h"
+#include "utils.h"
+#include "maps.h"
+#include "dbg.h"
+#include <stdint.h>
+#include <stdbool.h>
+
+
+/**
+ * update_data_metrics
+ * @direction:	0: Ingress 1: Egress
+ * @reason:	reason for forwarding or dropping packet.
+            	reason is 0 if packet is being forwarded, else reason
+            	is the drop error code.
+ * Update the data metrics map.
+ */
+static inline void update_data_metrics(__u32 bytes, __u8 direction, __u8 reason)
+{
+    struct data_metrics_value *entry, newEntry = {};
+    struct data_metrics_key key = {};
+
+    key.reason = reason;
+    key.dir     = direction;
+
+
+    if ((entry = map_lookup_elem(&cilium_metrics, &key))) {
+            __sync_fetch_and_add(&entry->count, 1);
+            __sync_fetch_and_add(&entry->bytes, bytes);
```

Comment:
Is there any performance impact from using sync_fetch_and_add here? How about using a per cpu map and aggregating when we pull the metrics out into golang side.

Reply:
Yes, we need to make this a per CPU map, and I will be raising a separate GH issue for that.
I am starting this groundwork off by the HASH_MAP to get it working end-to-end for 1.1

Reply:
Fair enough can you create the GH issue and link to it from the PR for this work just so we don't lose it.

Reply:
will do 👍

Reply:
Raised #4196 . Will put in in comments too

Reply:
done